lib/Sema/TypeCheckType.cpp

```diff
@@ -2634,7 +2633,7 @@ Type TypeResolver::resolveASTFunctionType(
   if (resolveASTFunctionTypeParams(repr->getArgsTypeRepr(), options,
                                    repr->getGenericEnvironment() != nullptr,
                                    diffKind, params)) {
-    return Type();
+    return ErrorType::get(Context);
```

Comment:
`resolveASTFunctionTypeParams` returns `Type()` if `resolveType` returns `Type()`. But after your changes, `resolveType` should never return `Null`, and you'll automatically have an error type in the parameter list, and that parameter list can just be used directly. You can refactor `resolveASTFunctionTypeParams` to return the parameters directly now.

Reply:
For this one, should resolveASTFunctionTypeParams return params instead of bool?

Reply:
params is a SmallVectorImpl<>. it looks like llvm wants that to be an out parameter not a return value (based on the documentation and on that it does not have a copy constructor). If I rewrite resolve to return a SmallVectorImpl it does not like it, unless I return a pointer, which it then has to delete, etc. Maybe just call resolveASTFunctionTypeParams and remove the check? I can replace     if (!ty) return true; with an assertion in resolveASTFunctionTypeParams. It would always return false then.

Reply:
Sure. So we can change the return type to a suitably-sized SmallVector of parameters.

Reply:
I just pushed a patch that changes the return type. I missed that SmallVector was the thing you return and that SmallVectorImpl can be an out parameter.